apps/contributions/tests/test_views.py

```diff
@@ -485,6 +485,47 @@ def test_contributor_call_excludes_payments_requiring_source(self, celery_task_m
         assert self.contribution_4.id not in contribution_ids
 
 
+class TestContributionsViewSetExportCSV(AbstractTestCase):
+    def setUp(self):
+        super().setUp()
+        self.set_up_domain_model()
+
+    def list_contributions(self, user):
+        self.client.force_authenticate(user=user)
+        return self.client.get(reverse("contribution-list"))
+
+    def email_contributions(self, user):
+        self.client.force_authenticate(user=user)
+        return self.client.post(reverse("contribution-email-contributions"))
+
+    def test_user_without_role(self):
+        self.generic_user.roleassignment = None
+        response = self.email_contributions(self.generic_user)
+        assert response.status_code == 403
+
+    @mock.patch("apps.contributions.views.send_templated_email_with_attachment.delay")
+    def test_user_with_role(self, email_mock):
+        response = self.email_contributions(self.rp_user)
+        assert response.status_code == 200
+
+        response = self.email_contributions(self.org_user)
+        assert response.status_code == 200
+
+        response = self.email_contributions(self.hub_user)
+        assert response.status_code == 200
+
+        response = self.email_contributions(self.contributor_user)
+        assert response.status_code == 403
+
+    @mock.patch("apps.contributions.views.send_templated_email_with_attachment.delay")
+    def test_data_in_csv_matching_with_contributions_list(self, email_mock):
+        contributions = self.list_contributions(self.rp_user).json()
+        self.email_contributions(self.rp_user).json()
+        contributions_csv = email_mock.call_args_list[0].kwargs["attachment"].split("\r\n")
+
+        assert len(contributions_csv) - 2 == contributions["count"]
+
```

Comment:
Seems strange to me to get the eligible contributions via view instead of by querying contributions table. We should be doing more to ensure not just count is correct, but that the right contributions appear. Also, our primary user these days is org admin instead of rp admin, so probably makes sense to make that user center of this test.

Something in this direction would be a better test, IMO (note that suggestion below assumes adding `id` to export as suggested below)

```suggestion
    def test_csv_data_is_correct_for_user(self, email_mock):
        expected_contribution_ids = Contribution.objects.filter(donation_page__revenue_program__in=self.org_user.roleassignment.revenue_programs.all()).values_list('id')
        assert len(expected_contribution_ids) >= 1
        self.email_contributions(self.rp_user)
        contributions_data = [row for row in csv.DictReader(email_mock.call_args_list[0].kwargs["attachment"].splitlines())]
      
        assert len(contributions_data) == expected_contribution_ids
        assert set([x['id'] for x in contributions_data]) == set(expected_contribution_ids)
```

Reply:
Comparing with the contributions endpoint because to check the AC `All data shown in contributions dashboard is included in export (same # line items)`.

Reply:
Addressed by 60502226689f39f2247fa04b815034cc91351356